test/integration/spec/rtcpeerconnection.js

```diff
@@ -786,43 +759,132 @@ function makeStream(constraints) {
+  beforeEach(async () => {
+    test = await makeTest();
+    tracks = getTracks(test.peerConnection);
   });
 
-  // NOTE(mroberts): See the comment in lib/webrtc/rtcpeerconnection/firefox.js
-  // for an explanation as to why we test this API the way we do.
-  it('should add each of the MediaStream\'s MediaStreamTracks to the RTCPeerConnection', () => {
-    test.peerConnection.addStream(stream);
-    if (isFirefox) {
-      const expectedTracks = stream.getTracks();
-      const actualTracks = test.peerConnection.getLocalStreams()[0].getTracks();
-      assertMediaStreamTracksEqual(actualTracks, expectedTracks);
-      return;
-    }
-    assert.equal(test.peerConnection.getLocalStreams()[0], stream);
-  });
+  [
+    [
+      'when the RTCPeerConnection is closed',
+      () => test.peerConnection.close()
+    ],
+    [
+      'when the MediaStreamTrack is already added to the RTCPeerConnection',
+      () => test.peerConnection.addTrack(trackToAdd, stream)
+    ]
+  ].forEach(([scenario, setup]) => {
+    context(scenario, () => {
+      var exception;
+
+      beforeEach(() => {
+        setup();
+        try {
+          test.peerConnection.addTrack(trackToAdd, stream);
+        } catch (e) {
+          exception = e;
+        }
+      });
 
-  context('when adding a stream that is already added', () => {
-    it('should not throw an exception', () => {
-      test.peerConnection.addStream(stream);
-      assert.doesNotThrow(() => test.peerConnection.addStream(stream));
+      it('should throw', () => {
+        assert(exception);
+      });
+
+      it('should not change the RTCPeerConnection\'s MediaStreamTracks', () => {
+        const tracksAfter = new Set(getTracks(test.peerConnection));
+        assert.deepEqual(tracksAfter, tracks);
+      });
+
+      afterEach(() => {
+        exception = null;
+      });
     });
+  });
+
+  it('should add each of the MediaStreamTracks to the RTCPeerConnection', () => {
+    const senders = addStream(test.peerConnection, stream);
+    const addedTracks = getTracks(test.peerConnection);
+    assert.deepEqual(addedTracks, stream.getTracks());
+    assert.deepEqual(senders.map(sender => sender.track), stream.getTracks());
+  });
+}
+
+function testRemoveTrack() {
+  var test;
+  var tracks;
+  var stream;
+  var localAudioSender;
+  var localAudioTrack;
+  var localVideoSender;
 
-    it('should not add the stream', () => {
-      test.peerConnection.addStream(stream);
-      test.peerConnection.addStream(stream);
-      assert.equal(test.peerConnection.getLocalStreams().length, 1);
+  before(async () => {
+    stream = await makeStream();
+    localAudioTrack = stream.getAudioTracks()[0];
+  });
+
+  beforeEach(async () => {
+    test = await makeTest();
+    const senders = addStream(test.peerConnection, stream);
+    localAudioSender = senders.find(sender => sender.track.kind === 'audio');
+    localVideoSender = senders.find(sender => sender.track.kind === 'video');
```

Comment:
In general, this is unsafe, since `track` may be `null`. Can we just compare on `track === localAudioTrack`, for example?

Reply:
I guess we can leverage the fact that we implemented `addStream` ourselves, and so these should be non-`null`, but it'd be nice if we could be more general.